src/array.jl

```diff
@@ -389,17 +389,18 @@ end
 # Methods preserving levels and more efficient than AbstractArray fallbacks
 copy(A::CategoricalArray) = deepcopy(A)
 
-function copy!(dest::CategoricalArray{T, N}, dstart::Integer,
-               src::CategoricalArray{T, N}, sstart::Integer,
-               n::Integer=length(src)-sstart+1) where {T, N}
+function copy!(dest::CategoricalArray{<:Union{T,Null}, N}, dstart::Integer,
+               src::S, sstart::Integer, n::Integer=length(src)-sstart+1) where
+               {T,N,S<:Union{CategoricalArray{<:Union{T,Null},N},SubArray{A,B,C,D}} where {A,B,C<:CategoricalArray{<:Union{T,Null},N},D}}
```

Comment:
You should be able to make this shorter by parameterizing `CA_OR_CA_VIEW` on the element type and using it here. Also avoid defining type parameters that are not used where possible, and note that the package currently uses spaces after commas even for type parameters.